crates/ruff_db/src/system/memory_fs.rs

```diff
@@ -237,6 +238,24 @@ impl MemoryFileSystem {
         let normalized = SystemPath::absolute(path, &self.inner.cwd);
         normalized.into_utf8_path_buf()
     }
+
+    pub(crate) fn read_dir(&self, path: impl AsRef<SystemPath>) -> Result<DirectoryIterator> {
+        let by_path = self.inner.by_path.read().unwrap();
+        let normalized = self.normalize_path(path.as_ref());
+        let entry = by_path.get(&normalized).ok_or_else(not_found)?;
+        if entry.is_file() {
+            return Err(not_a_directory());
+        };
+        // Is this inefficient? Yes, but this implementation of `System` is just for testing,
```

Comment:
I think you want to use `metadata` here to raise the correct error when the file doesn't exist:

```rust
        let metadata = self.metadata(path)?;
        if !metadata.file_type().is_directory() {
            return Err(not_a_directory());
        }
```

Reply:
Isn't that already taken care of a couple of lines above, where I do

```rs
let entry = by_path.get(&normalized).ok_or_else(not_found)?;
```

?